### DISCUSSION TITLE
variables

### DISCUSSION THREAD

**Discussion by NoorsTah** [elsa-workflows/elsa-core#5198] [Q&A]
is there a way to get variable from endpoint into custom activity as input ( with v2 ) switched because i need signals ?
While writing code (NO DESIGNER , NO JSON).
i'4ve been searching for two days now , no sample on v2.141 on this

**Comment by glime-ai** [comment]
To get a variable from an HTTP endpoint into a custom activity as input in Elsa 2.0, you can follow these steps:

1. Define an HTTP Endpoint activity that will receive the HTTP request and extract the variables from the request body.
2. Define your custom activity with an input property that will receive the variable from the HTTP Endpoint activity.
3. Connect the HTTP Endpoint activity to your custom activity in the workflow definition, ensuring that the output of the HTTP Endpoint is passed as an input to your custom activity.

Here's a simplified example of how you might define your custom activity with an input property:

```csharp
public class MyCustomActivity : Activity
{
    [ActivityInput(Hint = "The input value to process.")]
    public string MyInput { get; set; }

    protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
    {
        // Your custom logic here, using MyInput as needed.
        return Done();
    }
}
```

And here's how you might define your workflow in code:

```csharp
public class MyWorkflow : IWorkflow
{
    public void Build(IWorkflowBuilder builder)
    {
        builder
            .StartWith<HttpEndpoint>(
                setup => setup
                    .WithPath("/my-endpoint")
                    .WithMethod(HttpMethods.Post)
                    .WithReadContent())
            .Then<MyCustomActivity>(
                setup => setup
                    .WithInput(activity => activity.MyInput, context => context.GetInput<MyRequestModel>().MyProperty));
    }
}

public class MyRequestModel
{
    public string MyProperty { get; set; }
}
```

In this example, `HttpEndpoint` is an activity that listens for HTTP POST requests on `/my-endpoint`. It reads the content of the request and makes it available to the workflow context. The `MyCustomActivity` then takes an input property `MyInput`, which is set to the value of `MyProperty` from the request model.

When you send an HTTP POST request to `/my-endpoint` with a JSON body containing `MyProperty`, the `HttpEndpoint` activity will parse the request body into an instance of `MyRequestModel`, and the `MyCustomActivity` will receive the `MyProperty` value as its input.

Please note that this is a simplified example and you may need to adjust it to fit your specific requirements. Also, ensure that you have the necessary packages installed and configured for your project to support HTTP activities and custom activities.

**Comment by glime-ai** [comment]
Yes, you can get a variable from an endpoint into a custom activity as input in Elsa 2.x by using the `SetVariable` and `GetVariable` methods within your custom activity code. These methods allow you to read and write workflow variables, which can be used to pass data between activities.

Here's a simplified example of how you might define a custom activity that reads a variable:

```csharp
public class MyCustomActivity : Activity
{
    protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
    {
        // Read a variable from the workflow context
        var myVariable = context.GetVariable<string>("MyVariableName");

        // Additional logic using the variable...

        return Done();
    }
}
```

And here's how you might set a variable from an HTTP Endpoint:

```csharp
public class MyHttpEndpointActivity : Activity
{
    protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
    {
        // Assume you've obtained some data from the request
        var requestData = "some data from the request";

        // Set a variable in the workflow context
        context.SetVariable("MyVariableName", requestData);

        return Done();
    }
}
```

In the above example, `MyVariableName` is the key for the variable you're setting and getting. You would set this variable in one activity (like an HTTP Endpoint that receives data), and then you could access it in another activity (like your custom activity).

Please note that the exact implementation details may vary based on your specific use case and the version of Elsa you are using. Make sure to refer to the Elsa documentation and the source code for the version you are using to ensure compatibility.
